File: frontend/src/game/engine/board.ts
import { rngNext, type RngState } from "./prng";
import type { Board, Modifier } from "./types";

export function emptyCells(board: Board): number[] {
  const out: number[] = [];
  for (let i = 0; i < 16; i++) if (board[i] === 0) out.push(i);
  return out;
}

export function spawnTile(
  board: Board,
  rng: RngState,
  rngCallCount: number,
  modifier: Modifier
): { board: Board; rng: RngState; rngCallCount: number } {
  // Full-board guard: no empty cell -> spawn nothing, consume no RNG.
  // A Go/Gno port MUST mirror this early-return (Go would otherwise panic on % 0).
  const empties = emptyCells(board);
  if (empties.length === 0) {
    return { board: board.slice(), rng, rngCallCount };
  }
  // position draw
  let r = rngNext(rng);
  const pos = empties[r.value % empties.length];
  rng = r.state;
  // value draw
  r = rngNext(rng);
  let value = r.value % 10 === 0 ? 4 : 2;
  rng = r.state;
  if (modifier === "doubles") value *= 2;

  const next = board.slice();
  next[pos] = value;
  return { board: next, rng, rngCallCount: rngCallCount + 2 };
}

File: backend/internal/blockparty/engine/board.go
package engine

type Board = []int
type Modifier = string
type Move = string

func EmptyCells(b Board) []int {
	out := make([]int, 0, 16)
	for i := 0; i < 16; i++ {
		if b[i] == 0 {
			out = append(out, i)
		}
	}
	return out
}

// SpawnTile mirrors the TS spawnTile exactly, including the full-board guard:
// with no empty cell it spawns nothing and consumes no RNG (JS returns NaN on
// % 0; Go would panic — the guard makes both languages agree).
func SpawnTile(b Board, rng uint32, rngCallCount int, mod Modifier) (Board, uint32, int) {
	empties := EmptyCells(b)
	next := append(Board(nil), b...) // copy
	if len(empties) == 0 {
		return next, rng, rngCallCount
	}
	// position draw
	val, st := RngNext(rng)
	pos := empties[int(val%uint32(len(empties)))] // #nosec G115 -- len(empties) is 0..16, well within uint32
	rng = st
	// value draw
	val, st = RngNext(rng)
	v := 2
	if val%10 == 0 {
		v = 4
	}
	rng = st
	if mod == "doubles" {
		v *= 2
	}
	next[pos] = v
	return next, rng, rngCallCount + 2
}
